### PR TITLE
[PR] Streamlined port label placement

### PR DESCRIPTION
**Pull request by uruuru**
See #626. 

It probably makes sense to go through the PR commit by commit. 

There's one `TODO` open: where to implement the legacy behavior (see comment in code). 

While working on this (and in particular while writing the test), I found the behavior rather inconsistent: 
* inside and outside behave differently by default (inside even behaves differently for non- and hierarchical edges) 
* always same side and space efficient do not apply to inside labels 
* north/south behaves differently by default

I'd suggest to create another issue for this, though.

### CODE REVIEW

File: plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/NodeContext.java [hunk 1]
```diff
@@ -138,7 +138,8 @@ public NodeContext(final GraphAdapter<?> parentGraph, final NodeAdapter<?> node)
         sizeOptions = node.getProperty(CoreOptions.NODE_SIZE_OPTIONS);
         portConstraints = node.getProperty(CoreOptions.PORT_CONSTRAINTS);
         portLabelsPlacement = node.getProperty(CoreOptions.PORT_LABELS_PLACEMENT);
-        portLabelsNextToPort = node.getProperty(CoreOptions.PORT_LABELS_NEXT_TO_PORT_IF_POSSIBLE);
+        portLabelsNextToPort = portLabelsPlacement.contains(PortLabelPlacement.NEXT_TO_PORT_IF_POSSIBLE);
+
```
[the remark this answers is not in the document]
Comment (uruuru) [reply plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/NodeContext.java:142]: I agree, wanted to keep the changes small for this PR. I'll add this to another issue's description.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk [hunk 2]
```diff
@@ -725,22 +725,20 @@ group port {
...
-            "Decides on a placement method for port labels."
-        default = PortLabelPlacement.OUTSIDE
+            "Decides on a placement method for port labels; if empty, the node label's position is not
+            modified."
+        default = PortLabelPlacement.outside
         targets nodes
     }
 
-    option nextToPortIfPossible: boolean {
+    // TODO where to add the deprecated thingy? both graph import in layered and in common label placement?
```
[the remark this answers is not in the document]
Comment (uruuru) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/Core.melk]: I went for the visitor. Currently it's only applied by the recursive graph layout engine, i.e. not when calling the layout algorithm directly, but I could live with that.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/options/PortLabelPlacement.java [hunk 3]
```diff
@@ -9,19 +9,93 @@
...
+     * Cases in which this behavior is not the default:
+     * <ul>
+     *   <li>Outside labels of both hierarchical and non-hierarchical nodes,</li>
+     *   <li>Inside labels of hierarchical nodes.</li>
+     * </ul>
+     * 
+     * Cases in which the behavior is the default anyway:
+     * <ul>
+     *   <li>Inside labels of non-hierarchical nodes.</li>
+     * </ul>
```
[the remark this answers is not in the document]
Comment (uruuru) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/options/PortLabelPlacement.java:55]: I created #638.

File: plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/HorizontalPortPlacementSizeCalculator.java [hunk 4]
```diff
@@ -259,8 +259,8 @@ private static void calculateHorizontalNodeSizeRequiredByFreePorts(final NodeCon
     private static void setupPortMargins(final NodeContext nodeContext, final PortSide portSide) {
         Collection<PortContext> portContexts = nodeContext.portContexts.get(portSide);
         
-        boolean portLabelsOutside = nodeContext.portLabelsPlacement == PortLabelPlacement.OUTSIDE;
-        boolean spaceEfficientPortLabels = nodeContext.sizeOptions.contains(SizeOptions.SPACE_EFFICIENT_PORT_LABELS)
+        boolean portLabelsOutside = nodeContext.portLabelsPlacement.contains(PortLabelPlacement.OUTSIDE);
+        boolean spaceEfficientPortLabels = nodeContext.portLabelsPlacement.contains(PortLabelPlacement.SPACE_EFFICIENT)
```
[the remark this answers is not in the document]
Comment (uruuru) [reply plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/HorizontalPortPlacementSizeCalculator.java]: 👍, part of the graph visitor.

File: plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeLabelAndSizeUtilities.java [hunk 5]
```diff
@@ -204,8 +204,11 @@ public static boolean areSizeConstraintsFixed(final NodeContext nodeContext) {
      * Outside ports usually have their labels placed below or to the right. The first port, however, may have its label
      * placed on the other side. This is true if several conditions apply:
      * <ul>
-     *   <li>The port actually wants its label to be placed next to it.</li>
-     *   <li>There are only two ports or space-efficient port label placement is turned on.</li>
+     *   <li>The port does <b>not</b> want its label to be placed next to it (see
+     *       {@link PortLabelPlacement#NEXT_TO_PORT_IF_POSSIBLE}.</li>
```
[the remark this answers is not in the document]
Comment (uruuru) [reply plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeLabelAndSizeUtilities.java:208]: In that case the method would have to check for an incident edge, right? 

I think I didn't change the implementation but only adjusted the comment.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/options/PortLabelPlacement.java [hunk 6]
```diff
@@ -9,19 +9,93 @@
...
+
 /**
- * Definition of port label placement strategies. The chosen strategy determines whether port
- * labels are placed inside or outside of the respective node.
- *
- * @author jjc
+ * Options for controlling how port labels are placed by layout algorithms. The corresponding layout
+ * option will usually accept an {@link EnumSet} over this enumeration, theoretically allowing
+ * arbitrary and even contradictory subsets of options to be set. Note that you are restricted to
+ * the following combinations if you want your choice to make sense:
```
[the remark this answers is not in the document]
Comment (uruuru) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/options/PortLabelPlacement.java:21]: I added an 'isValid'  method to the enum which is checked when retrieving the option in `NodeContext`. Any suggestion for a better/other place to check it?

While at it I also added it for `NodeLabelPlacement` as well. The change may result in graphs failing that have been working before since so far we failed silently. However, I feel it's better to let the user know the configuration is off in this case.

[the remark this answers is not in the document]
Comment (le-cds) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/options/PortLabelPlacement.java:21]: This is a bit of a philosophical question. Ideally, I'd vote for failing as early as possible. What that means is up for debate, though.

At one extreme, one would add additional `IGraphElementVisitor`s to check for configuration errors. Pro: we'd fail even before a layout algorithm is run. Contra: we would either need a great deal of visitors and some sort of registry mechanism, or we'd have a couple of large visitors that combine concerns from very different parts of our code, making everything harder to maintain.

The other extreme is checking for configuration errors as soon as they become relevant, which is what you've done here. Given that this keeps concerns separated and that it's easier to maintain, I think this is the way to go.

tl;dr: well done. 😄

File: plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeLabelAndSizeUtilities.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (le-cds) [reply plugins/org.eclipse.elk.alg.common/src/org/eclipse/elk/alg/common/nodespacing/internal/algorithm/NodeLabelAndSizeUtilities.java:208]: You're right. The problem was my poor choice of words in the original comment. By "next to port" I meant a label placed above or below its port (taking an eastern port as an example here). Usually, however, we understand the term to mean a label placed right next to the port, where the edge would usually be routed.

I'm still wondering what happens if an edge is in the way. I guess in this code we simply accept that this is the case and that the user will know what they're doing. This is different to how inside ports behave, which is what #638 is about.
